numpy/lib/arraypad.py

```diff
@@ -1165,191 +843,133 @@ def pad(array, pad_width, mode, **kwargs):
            [100, 100, 100, 100, 100, 100, 100],
            [100, 100, 100, 100, 100, 100, 100]])
     """
-    if not np.asarray(pad_width).dtype.kind == 'i':
+    array = np.asarray(array)
+    pad_width = np.asarray(pad_width)
```

Comment:
Sorry, nvm. this is an array input. I think this check is nicer in `_as_pairs` though probably, since `as_pairs` and `as_pairs` has an unnecessary round right now (probably some left over from old code).

Reply:
I see what you mean but its not that easy. Moving the `dtype.kind == 'i'` check into `_as_pairs` would require another keyword telling `_as_pairs` when to convert (see `as_index` parameter) and when to raise an exception. I think that's the more complicated option. And it would only be used once anyway.

Or do I misunderstand you intention? Which round is unnecessary? The requirement for how `_as_pairs` behaves was quite strict in order to fit into the old implementation (see #11966). A good example for some of the disadvantages of the incremental approach. ;) However there may be some streamlining possible now. I'll think on it.

Reply:
Well, arguably, the `stat_length` input should also array when it is not an integer, I did not see it actually used that round there.

Arguably, that should should get a DeprecationWarning and be removed in the long run. But I am OK with leaving that for later. If you keep it as is, it does not matter much.